src/cli/telemetry/telemetry.ts

```diff
@@ -0,0 +1,203 @@
+import { readConfig, updateConfig } from '../ionic-config';
+import { isObject } from '@utils';
+
+export interface CommandMetricData {
+	// app_id: string;
+	command: string;
+	arguments: string;
+	options: string;
+	duration: number;
+	error: string | null;
+	node_version: string;
+	// os: string;
+}
+
+export interface Metric<N extends string, D> {
+	name: N;
+	timestamp: string;
+	source: 'stencil_cli';
+	value: D;
+}
+
+// Used for deeper understanding on how people are using certain output targets
+interface OutputTargetOptions {
+	serviceWorker?: boolean;
+}
+
+type TelemetryCallback = (...args: any[]) => void | Promise<void>;
+
+export interface TrackableData {
+	command?: string;
+	token?: string;
+	arguments?: string[];
+	targets?: string[][] | OutputTargetOptions[][];
+	task?: string;
+	duration?: number;
+	packages?: string[][];
+	telemetry?: boolean;
+	ci?: boolean;
+	platform: string;
+	typescript: string;
+	rollup: string;
+	compiler: string;
+	system: string;
+	build: string;
+	stencil: string;
+}
+
+export interface telemetryActionArgs {
+	args?: string[],
+	sys?: CompilerSystem,
+	logger?: Logger,
+	config?: Config,
+	action?: TelemetryCallback
+}
+
+export async function telemetryAction(options: telemetryActionArgs) {
+	const { args, sys, logger, config, action } = options;
+	const coreCompiler = await loadCoreCompiler(sys);
+	const details = sys.details;
+	const versions = coreCompiler.versions;
+	const flags = parseFlags(args, sys);
+	let duration = undefined;
+	let error: any;
+
+	if (action) {
+		const start = new Date();
+
+		try {
+			await action();
+		} catch (e) {
+			error = e;
+		}
+
+		const end = new Date();
+		duration = end.getTime() - start.getTime();
+	}
+	
+	const token = await getTelemetryToken();
+	const packages = await getInstalledPackages();
+	const targets = await getActiveTargets(config);
+
+	const data: TrackableData = {
+		token,
+		duration,
+		targets,
+		packages: packages || [],
+		arguments: flags.args,
+		ci: !!flags.ci,
+		task: flags.task,
+		stencil: coreCompiler.version,
+		system: `${sys.name} ${sys.version}`,
+		platform: `${details.platform} (${details.release})`,
+		compiler: sys.getCompilerExecutingPath(),
+		build: coreCompiler.buildId,
+		typescript: versions.typescript,
+		rollup: versions.rollup
+	};
+
+	if (await shouldTrack(data.ci)) {
+		await sendMetric('stencil_cli_command', data);
+		logger.debug(`${logger.blue('Telemetry')}: ${logger.gray(JSON.stringify(data))}`)
+	}
+
+	if (error) {
+		throw error;
+	}
+}
+
+async function getActiveTargets(config: Config): Promise<string[][] | OutputTargetOptions[][]> {
+	return (config.outputTargets as any[]).map(target => {
+		let outputTargetSettings: OutputTargetOptions = {};
+
+		if (target.type === "www") {
+			outputTargetSettings.serviceWorker = isObject(target?.serviceWorker)
+		}
+
+		return [target.type, outputTargetSettings]
+	});
+}
+
+async function getInstalledPackages() {
+	// Read package.json
```

Comment:
Rather than reading the package.json it might be more useful to determine which versions are actually installed. You can get this info via e.g. `npm ls @stencil/core`, but there may be a better/more efficient way.

Reply:
I took a separate approach, let me know what you think.